# How to Start connecting spark using jupyter

# Creating Spark Context Connections

In [1]:
#importing sparkContext
from pyspark import SparkContext

In [2]:
# Create or use existing sparkcontext
sc = SparkContext.getOrCreate()

In [3]:
#Print object SC
print (sc)

<SparkContext master=local[*] appName=PySparkShell>


In [4]:
#Print object SC
print (sc.version)

2.2.0


In [5]:
from pyspark import SparkContext
sc = SparkContext.getOrCreate()
print (sc)

<SparkContext master=local[*] appName=PySparkShell>


In [6]:
from pyspark import SparkContext
sc = SparkContext.getOrCreate()
print (sc.version)

2.2.0


# Creating Spark Session for SQL


In [7]:

import pyspark

In [8]:
# We will import Spark session library
from pyspark.sql  import SparkSession

In [9]:
# Start Spark session and give appname of your choice
spark = SparkSession.builder.appName('Basics').getOrCreate()
# it might take longer, have patience unless there is an error

In [10]:
print(spark)

In [11]:
print (spark.version)

2.2.0


In [12]:
# Merged all statement

In [13]:
#Create Spark Session
import pyspark
from pyspark.sql  import SparkSession
spark = SparkSession.builder.appName('Basics').getOrCreate()
print(spark)

In [14]:
# Merged all statement

In [15]:
#Create Spark Session
import pyspark
from pyspark.sql  import SparkSession
spark = SparkSession.builder.appName('Basics').getOrCreate()
print(spark.version)

2.2.0


# 2

# Dataframe Basics and Json

Lets create Json data first.
Java Script Object Notation
Create a file with extension "i am using emp.json"
Copy below data in file

{"name":"Deepak"}
{"name":"Amit"}
{"name":"Rudra", "age":20}
{"name":"Parul", "age":21}
{"name":"Sikha", "age": 20,"salary":200}

In [2]:
# Create connection
import pyspark
from pyspark.sql  import SparkSession
spark = SparkSession.builder.appName('Algae').getOrCreate()
print(spark)

In [3]:
# read Josn file
df = spark.read.json('C:\\sparkinstall\\spark-2.2.0-bin-hadoop2.7\\bin\\DATASET_FILES\\emp1.json')
print(df)

DataFrame[age: double, name: string, salary: bigint]


In [4]:
#Print data from file
df.show()
#{age:NULL , Name: Deepak, salary:NULL}
#{ Name: Deepak}

+----+------+------+
| age|  name|salary|
+----+------+------+
|null|Deepak|  null|
|null|  Amit|  null|
|21.5| Rudra|  null|
|21.0| Parul|  null|
|20.5| Sikha|   200|
+----+------+------+



In [5]:
df.collect()

[Row(age=None, name='Deepak', salary=None),
 Row(age=None, name='Amit', salary=None),
 Row(age=21.5, name='Rudra', salary=None),
 Row(age=21.0, name='Parul', salary=None),
 Row(age=20.5, name='Sikha', salary=200)]

In [6]:
df.head(2)

[Row(age=None, name='Deepak', salary=None),
 Row(age=None, name='Amit', salary=None)]

In [7]:
df.select(['age','name'])

DataFrame[age: double, name: string]

In [8]:
df.select(['age','name']).show()

+----+------+
| age|  name|
+----+------+
|null|Deepak|
|null|  Amit|
|21.5| Rudra|
|21.0| Parul|
|20.5| Sikha|
+----+------+



In [9]:
df.select('*').show()

+----+------+------+
| age|  name|salary|
+----+------+------+
|null|Deepak|  null|
|null|  Amit|  null|
|21.5| Rudra|  null|
|21.0| Parul|  null|
|20.5| Sikha|   200|
+----+------+------+



 # Adding a new column for projection only - withColumn

In [10]:
# Adding a new column  for projection only
df.withColumn('age1',df['age']).show()

+----+------+------+----+
| age|  name|salary|age1|
+----+------+------+----+
|null|Deepak|  null|null|
|null|  Amit|  null|null|
|21.5| Rudra|  null|21.5|
|21.0| Parul|  null|21.0|
|20.5| Sikha|   200|20.5|
+----+------+------+----+



In [11]:
df.withColumn('age+',df['age']+1).show()

+----+------+------+----+
| age|  name|salary|age+|
+----+------+------+----+
|null|Deepak|  null|null|
|null|  Amit|  null|null|
|21.5| Rudra|  null|22.5|
|21.0| Parul|  null|22.0|
|20.5| Sikha|   200|21.5|
+----+------+------+----+



In [12]:
df.withColumn('age*',df['age']*2).show()

+----+------+------+----+
| age|  name|salary|age*|
+----+------+------+----+
|null|Deepak|  null|null|
|null|  Amit|  null|null|
|21.5| Rudra|  null|43.0|
|21.0| Parul|  null|42.0|
|20.5| Sikha|   200|41.0|
+----+------+------+----+



In [13]:
df.withColumn('divide_age',df['age']/2).show()

+----+------+------+----------+
| age|  name|salary|divide_age|
+----+------+------+----------+
|null|Deepak|  null|      null|
|null|  Amit|  null|      null|
|21.5| Rudra|  null|     10.75|
|21.0| Parul|  null|      10.5|
|20.5| Sikha|   200|     10.25|
+----+------+------+----------+



In [14]:
# Dataframe
df.withColumn('divide_age',df['age']/2)

DataFrame[age: double, name: string, salary: bigint, divide_age: double]

# Reading Structure of Data frame¶

In [15]:
# print Schema\structure of dataframe along nullability
df.printSchema()

root
 |-- age: double (nullable = true)
 |-- name: string (nullable = true)
 |-- salary: long (nullable = true)



In [18]:
# print list of columns
df.columns

['age', 'name', 'salary']

In [19]:
# Describe the structure of dataframe
df.describe()

DataFrame[summary: string, age: string, name: string, salary: string]

In [20]:
df['age']

Column<b'age'>

In [21]:
type(df['age'])

pyspark.sql.column.Column

In [22]:
df.select('age')

DataFrame[age: double]

In [23]:
df.select('age').show()

+----+
| age|
+----+
|null|
|null|
|21.5|
|21.0|
|20.5|
+----+



In [24]:
type(df.select('age'))

pyspark.sql.dataframe.DataFrame

# 3

# SQL Query in PySpark¶
Lets create Json data first. Create a file with extension "i am using emp.json" Copy below data in file

{"name":"Deepak"} {"name":"Amit"} {"name":"Rudra", "age":20} {"name":"Parul", "age":21} {"name":"Sikha", "age": 20,"salary":200}

In [25]:
# Create connection
import pyspark
from pyspark.sql  import SparkSession
spark = SparkSession.builder.appName('Algae').getOrCreate()
print(spark)

In [28]:
df = spark.read.json('C:\\sparkinstall\\spark-2.2.0-bin-hadoop2.7\\bin\\DATASET_FILES\\emp1.json')
df.show()

+----+------+------+
| age|  name|salary|
+----+------+------+
|null|Deepak|  null|
|null|  Amit|  null|
|21.5| Rudra|  null|
|21.0| Parul|  null|
|20.5| Sikha|   200|
+----+------+------+



# Register SQL View¶

In [29]:
# Register the DataFrame as a SQL temporary view
df.createOrReplaceTempView("emp")

In [30]:
sql_results = spark.sql("SELECT * FROM emp")

In [31]:
sql_results

DataFrame[age: double, name: string, salary: bigint]

In [32]:
sql_results.show()

+----+------+------+
| age|  name|salary|
+----+------+------+
|null|Deepak|  null|
|null|  Amit|  null|
|21.5| Rudra|  null|
|21.0| Parul|  null|
|20.5| Sikha|   200|
+----+------+------+



In [33]:
sql_results.groupBy("name").mean().show()

+------+--------+-----------+
|  name|avg(age)|avg(salary)|
+------+--------+-----------+
| Rudra|    21.5|       null|
| Sikha|    20.5|      200.0|
|Deepak|    null|       null|
|  Amit|    null|       null|
| Parul|    21.0|       null|
+------+--------+-----------+



# SQL queries on PYSQL View

In [34]:
spark.sql("SELECT * FROM emp WHERE age=21").show()

+----+-----+------+
| age| name|salary|
+----+-----+------+
|21.0|Parul|  null|
+----+-----+------+



In [35]:
spark.sql("SELECT * FROM emp WHERE age=21 or salary =200").show()

+----+-----+------+
| age| name|salary|
+----+-----+------+
|21.0|Parul|  null|
|20.5|Sikha|   200|
+----+-----+------+



In [36]:
sql_results = spark.sql("SELECT age, avg(salary) FROM emp  group by age")
sql_results.show()

+----+-----------+
| age|avg(salary)|
+----+-----------+
|20.5|      200.0|
|null|       null|
|21.0|       null|
|21.5|       null|
+----+-----------+



# 5 

# Connect Python to SQL Server¶

In [37]:
!Pip install pyodbc


In [ ]:
/***************Pre-requisite****************************/
Make sure to install pyodbc
Pip install pyodbc


***Run these steps directly on sql server*/

Create Database Algae GO Use Algae GO Create table emp (name varchar (100), age int) Go Insert into Emp values ('Ravi', 26), ('Nandini', 23), ('Nikita',24)

# Syntax To Create Connection
import pyodbc

conn = pyodbc.connect('Driver={SQL Server};' 'Server=Please_Update_Actual_SQL_Server_name;' 'Database=AdventureWorks2012;' 'Trusted_Connection=yes;')

In [38]:
import pyodbc

conn = pyodbc.connect('Driver={SQL Server};'
                      'Server=PTBLR11412\SQL2019;'
                      'Database=Algae;'
                      'Trusted_Connection=yes;')

OperationalError: ('08001', '[08001] [Microsoft][ODBC SQL Server Driver][DBNETLIB]SQL Server does not exist or access denied. (17) (SQLDriverConnect); [08001] [Microsoft][ODBC SQL Server Driver][DBNETLIB]ConnectionOpen (Connect()). (53)')

In [ ]:
cursor = conn.cursor()
cursor.execute('SELECT * FROM [Algae].[dbo].[emp]')

In [ ]:
for row in cursor:
    print(row)

In [ ]:
# Create connection
import pyspark
from pyspark.sql  import SparkSession
spark = SparkSession.builder.appName('Algae').getOrCreate()
print(spark)

# Pull SQL Data in Pandas Dataframe

In [ ]:
import pandas as pd
query = 'SELECT * FROM [Algae].[dbo].[emp]'
pdf = pd.read_sql(query, conn)

In [ ]:
print(pdf)

# Convert pandas dataframe to Spark Data frame

In [ ]:
sparkDF =  spark.createDataFrame(pdf)
sparkDF.show()

# 6 

# GroupBy and Aggregate Functions

In [39]:
# Create connection
import pyspark
from pyspark.sql  import SparkSession
spark = SparkSession.builder.appName('Algae').getOrCreate()
print(spark)

In [40]:
df = spark.read.csv('C:\\sparkinstall\\spark-2.2.0-bin-hadoop2.7\\bin\\DATASET_FILES\\sales_info.csv',inferSchema=True,header=True)

In [41]:
df.printSchema()

root
 |-- Company: string (nullable = true)
 |-- Person: string (nullable = true)
 |-- Sales: double (nullable = true)



In [43]:
df.show()

+-------+-------+-----+
|Company| Person|Sales|
+-------+-------+-----+
|   GOOG|    Sam|200.0|
|   GOOG|Charlie|120.0|
|   GOOG|  Frank|340.0|
|   MSFT|   Tina|600.0|
|   MSFT|    Amy|124.0|
|   MSFT|Vanessa|243.0|
|     FB|   Carl|870.0|
|     FB|  Sarah|350.0|
|   APPL|   John|250.0|
|   APPL|  Linda|130.0|
|   APPL|   Mike|750.0|
|   APPL|  Chris|350.0|
+-------+-------+-----+



Let's group together by company!

In [44]:
df.groupBy("Company")

## Syntax
## df.groupBy("Groupled Columns"). Any Aggregate fucntion

## This returns a GroupedData object, off of which you can all various methods

In [45]:
# Mean
df.groupBy("Company").mean().show()

+-------+-----------------+
|Company|       avg(Sales)|
+-------+-----------------+
|   APPL|            370.0|
|   GOOG|            220.0|
|     FB|            610.0|
|   MSFT|322.3333333333333|
+-------+-----------------+



In [46]:
# Count
df.groupBy("Company").count().show()

+-------+-----+
|Company|count|
+-------+-----+
|   APPL|    4|
|   GOOG|    3|
|     FB|    2|
|   MSFT|    3|
+-------+-----+



In [47]:
# Max
df.groupBy("Company").max().show()

+-------+----------+
|Company|max(Sales)|
+-------+----------+
|   APPL|     750.0|
|   GOOG|     340.0|
|     FB|     870.0|
|   MSFT|     600.0|
+-------+----------+



In [48]:
# Min
df.groupBy("Company").min().show()

+-------+----------+
|Company|min(Sales)|
+-------+----------+
|   APPL|     130.0|
|   GOOG|     120.0|
|     FB|     350.0|
|   MSFT|     124.0|
+-------+----------+



In [49]:
# Sum
df.groupBy("Company").sum().show()

+-------+----------+
|Company|sum(Sales)|
+-------+----------+
|   APPL|    1480.0|
|   GOOG|     660.0|
|     FB|    1220.0|
|   MSFT|     967.0|
+-------+----------+



Check out this link for more info on other methods: http://spark.apache.org/docs/latest/api/python/pyspark.sql.html#pyspark-sql-module

Not all methods need a groupby call, instead you can just call the generalized .agg() method, that will call the aggregate across all rows in the dataframe column specified. It can take in arguments as a single column, or create multiple aggregate calls all at once using dictionary notation.

For example:

In [50]:
# Max sales across everything
df.agg({'Sales':'max'}).show()

+----------+
|max(Sales)|
+----------+
|     870.0|
+----------+



In [51]:
# Could have done this on the group by object as well:

In [52]:
grouped = df.groupBy("Company")

In [53]:
grouped.agg({"Sales":'max'}).show()

+-------+----------+
|Company|max(Sales)|
+-------+----------+
|   APPL|     750.0|
|   GOOG|     340.0|
|     FB|     870.0|
|   MSFT|     600.0|
+-------+----------+



Functions
There are a variety of functions you can import from pyspark.sql.functions. Check out the documentation for the full list available: http://spark.apache.org/docs/latest/api/python/pyspark.sql.html#module-pyspark.sql.functions

In [54]:
from pyspark.sql.functions import countDistinct, avg,stddev

In [55]:
df.select(countDistinct("Sales")).show()

+---------------------+
|count(DISTINCT Sales)|
+---------------------+
|                   11|
+---------------------+



In [56]:
df.select(countDistinct("Sales").alias("Distinct Sales")).show()

+--------------+
|Distinct Sales|
+--------------+
|            11|
+--------------+



In [57]:
df.select(avg('Sales')).show()

+-----------------+
|       avg(Sales)|
+-----------------+
|360.5833333333333|
+-----------------+



In [58]:
df.select(stddev("Sales")).show()

+------------------+
|stddev_samp(Sales)|
+------------------+
|250.08742410799007|
+------------------+



# That is a lot of precision for digits! Let's use the format_number to fix that!

In [59]:
from pyspark.sql.functions import format_number

In [60]:
sales_std = df.select(stddev("Sales").alias('std'))

In [61]:
sales_std.show()

+------------------+
|               std|
+------------------+
|250.08742410799007|
+------------------+



In [62]:
# format_number("col_name",decimal places)
sales_std.select(format_number('std',2)).show()

+---------------------+
|format_number(std, 2)|
+---------------------+
|               250.09|
+---------------------+



# Order By
# You can easily sort with the orderBy method:

In [63]:
# OrderBy
# Ascending
df.orderBy("Sales").show()

+-------+-------+-----+
|Company| Person|Sales|
+-------+-------+-----+
|   GOOG|Charlie|120.0|
|   MSFT|    Amy|124.0|
|   APPL|  Linda|130.0|
|   GOOG|    Sam|200.0|
|   MSFT|Vanessa|243.0|
|   APPL|   John|250.0|
|   GOOG|  Frank|340.0|
|     FB|  Sarah|350.0|
|   APPL|  Chris|350.0|
|   MSFT|   Tina|600.0|
|   APPL|   Mike|750.0|
|     FB|   Carl|870.0|
+-------+-------+-----+



In [64]:
# Descending call off the column itself.
df.orderBy(df["Sales"].desc()).show()

+-------+-------+-----+
|Company| Person|Sales|
+-------+-------+-----+
|     FB|   Carl|870.0|
|   APPL|   Mike|750.0|
|   MSFT|   Tina|600.0|
|     FB|  Sarah|350.0|
|   APPL|  Chris|350.0|
|   GOOG|  Frank|340.0|
|   APPL|   John|250.0|
|   MSFT|Vanessa|243.0|
|   GOOG|    Sam|200.0|
|   APPL|  Linda|130.0|
|   MSFT|    Amy|124.0|
|   GOOG|Charlie|120.0|
+-------+-------+-----+



Most basic functions you would expect to be available are, so make sure to check out the documentation!



# 7

# Convert Data Type in PySpark
# Lets create Json data first. Create a file with extension "i am using emp.json" Copy below data in file

{"name":"Deepak"} {"name":"Amit"} {"name":"Rudra", "age":20} {"name":"Parul", "age":21} {"name":"Sikha", "age": 20,"salary":200}

In [65]:
# Create connection
import pyspark
from pyspark.sql  import SparkSession
spark = SparkSession.builder.appName('Algae').getOrCreate()
print(spark)

In [67]:
# read Josn file
df = spark.read.json('C:\\sparkinstall\\spark-2.2.0-bin-hadoop2.7\\bin\\DATASET_FILES\\emp1.json')
df.show()

+----+------+------+
| age|  name|salary|
+----+------+------+
|null|Deepak|  null|
|null|  Amit|  null|
|21.5| Rudra|  null|
|21.0| Parul|  null|
|20.5| Sikha|   200|
+----+------+------+



In [68]:
df.printSchema()

root
 |-- age: double (nullable = true)
 |-- name: string (nullable = true)
 |-- salary: long (nullable = true)



In [69]:
from pyspark.sql.types import StructField,StringType,IntegerType,StructType

Next we need to create the list of Structure fields

:param name: string, name of the field.

:param dataType: :class:DataType of the field
.
:param nullable: boolean, whether the field can be null (None) or not.

In [70]:
data_schema = [StructField("age", IntegerType(), True),StructField("name", StringType(), True)]

In [71]:
final_struc = StructType(fields=data_schema)

In [73]:
df = spark.read.json('C:\\sparkinstall\\spark-2.2.0-bin-hadoop2.7\\bin\\DATASET_FILES\\emp1.json', schema=final_struc)

In [74]:
df.printSchema()

root
 |-- age: integer (nullable = true)
 |-- name: string (nullable = true)



In [75]:
df.show()

+----+------+
| age|  name|
+----+------+
|null|Deepak|
|null|  Amit|
|null|  null|
|  21| Parul|
|null|  null|
+----+------+



# `Another Example

Lets create Json data first. Create a file with extension "i am using emp.json" Copy below data in file

{"name":"Deepak"} {"name":"Amit"} {"name":"Rudra", "age":21.5} {"name":"Parul", "age":21.5} {"name":"Sikha", "age": 20.5,"salary":200}

In [77]:
# read Josn file
df = spark.read.json('C:\\sparkinstall\\spark-2.2.0-bin-hadoop2.7\\bin\\DATASET_FILES\\emp1.json')
df.show()

+----+------+------+
| age|  name|salary|
+----+------+------+
|null|Deepak|  null|
|null|  Amit|  null|
|21.5| Rudra|  null|
|21.0| Parul|  null|
|20.5| Sikha|   200|
+----+------+------+



In [78]:
df.printSchema()

root
 |-- age: double (nullable = true)
 |-- name: string (nullable = true)
 |-- salary: long (nullable = true)



In [79]:
from pyspark.sql.types import StructField,StringType,IntegerType,StructType

In [80]:
data_schema = [StructField("age", IntegerType(), True),StructField("name", StringType(), True)]

In [81]:
final_struc = StructType(fields=data_schema)

In [83]:
df = spark.read.json('C:\\sparkinstall\\spark-2.2.0-bin-hadoop2.7\\bin\\DATASET_FILES\\emp1.json', schema=final_struc)

In [84]:
df.printSchema()

root
 |-- age: integer (nullable = true)
 |-- name: string (nullable = true)



In [85]:
df.show()

+----+------+
| age|  name|
+----+------+
|null|Deepak|
|null|  Amit|
|null|  null|
|  21| Parul|
|null|  null|
+----+------+

